vic-machine panics

```
[vagrant@devbox:/opt/go/src/github.com/vmware/vic(1578)] /opt/go/src/github.com/vmware/vic/infra/machines/vcsa/install.sh -a -i
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x5a3b08]

goroutine 1 [running]:
panic(0xb94260, 0xc4200100d0)
        /usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/vmware/vic/pkg/flags.(*optionalString).String(0xc4200260e0, 0xc21f40, 0xc4200260e0)
        /opt/go/src/github.com/vmware/vic/pkg/flags/optional_string.go:38 +0x8
flag.isZeroValue(0xc420019340, 0xd3fc7b, 0x5, 0xd53045)
        /usr/local/go/src/flag/flag.go:393 +0x12d
flag.(*FlagSet).PrintDefaults.func1(0xc420019340)
        /usr/local/go/src/flag/flag.go:467 +0x1d5
flag.(*FlagSet).VisitAll(0xc420014ae0, 0xc420153160)
        /usr/local/go/src/flag/flag.go:323 +0x67
flag.(*FlagSet).PrintDefaults(0xc420014ae0)
        /usr/local/go/src/flag/flag.go:476 +0x46
flag.defaultUsage(0xc420014ae0)
        /usr/local/go/src/flag/flag.go:509 +0x96
flag.(*FlagSet).usage(0xc420014ae0)
        /usr/local/go/src/flag/flag.go:825 +0x55
flag.(*FlagSet).failf(0xc420014ae0, 0xd69cab, 0x22, 0xc420153368, 0x1, 0x1, 0x18, 0x0)
        /usr/local/go/src/flag/flag.go:814 +0x116
flag.(*FlagSet).parseOne(0xc420014ae0, 0x1, 0x0, 0x0)
        /usr/local/go/src/flag/flag.go:873 +0xad8
flag.(*FlagSet).Parse(0xc420014ae0, 0xc4200016e0, 0x18, 0x18, 0x1576d38, 0x0)
        /usr/local/go/src/flag/flag.go:915 +0x60
github.com/vmware/vic/vendor/github.com/urfave/cli.Command.Run(0xd4052a, 0x6, 0x0, 0x0, 0x0, 0x0, 0x0, 0xd42125, 0xa, 0x0, ...)
        /opt/go/src/github.com/vmware/vic/vendor/github.com/urfave/cli/command.go:120 +0x69c
github.com/vmware/vic/vendor/github.com/urfave/cli.(*App).Run(0xc420192000, 0xc4200016c0, 0x1a, 0x1a, 0x0, 0x0)
        /opt/go/src/github.com/vmware/vic/vendor/github.com/urfave/cli/app.go:236 +0x60c
main.main()
        /opt/go/src/github.com/vmware/vic/cmd/vic-machine/main.go:98 +0x1756
[vagrant@devbox:/opt/go/src/github.com/vmware/vic(1578)]
```

following is how that script uses vic-machine
```
        /opt/go/src/github.com/vmware/vic/bin/vic-machine-linux create \
            -t ${host} \
            -u ${user} \
            -p ${pass} \
            -n ZzZ \
            -r /dc1/host/cluster1/Resources \
            --image-datastore vsanDatastore \
            --volume-store vsanDatastore:default \
            --external-network /dc1/network/ExternalNetwork \
            --bridge-network /dc1/network/InternalNetwork \
            --container-network /dc1/network/ExternalNetwork:external \
            --timeout 10m \
            --no-tls \
            -f
```

same result, this time I tried to deploy to ESXi

```
[vagrant@devbox:/opt/go/src/github.com/vmware/vic(master)] sudo make most
...
ISO image produced: 31342 sectors
Written to medium : 31520 sectors at LBA 32
Writing to '/opt/go/src/github.com/vmware/vic/bin/bootstrap.iso' completed successfully.

building vic-machine linux...
[vagrant@devbox:/opt/go/src/github.com/vmware/vic(master)] /opt/go/src/github.com/vmware/vic/infra/machines/vcsa/install.sh -e -i
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x5a3b08]

goroutine 1 [running]:
panic(0xb95100, 0xc4200120d0)
        /usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/vmware/vic/pkg/flags.(*optionalString).String(0xc4200280e0, 0xc22de0, 0xc4200280e0)
        /opt/go/src/github.com/vmware/vic/pkg/flags/optional_string.go:38 +0x8
flag.isZeroValue(0xc42001b340, 0xd40b9b, 0x5, 0xd53f73)
        /usr/local/go/src/flag/flag.go:393 +0x12d
flag.(*FlagSet).PrintDefaults.func1(0xc42001b340)
        /usr/local/go/src/flag/flag.go:467 +0x1d5
flag.(*FlagSet).VisitAll(0xc420016ae0, 0xc420155160)
        /usr/local/go/src/flag/flag.go:323 +0x67
flag.(*FlagSet).PrintDefaults(0xc420016ae0)
        /usr/local/go/src/flag/flag.go:476 +0x46
flag.defaultUsage(0xc420016ae0)
        /usr/local/go/src/flag/flag.go:509 +0x96
flag.(*FlagSet).usage(0xc420016ae0)
        /usr/local/go/src/flag/flag.go:825 +0x55
flag.(*FlagSet).failf(0xc420016ae0, 0xd6abd9, 0x22, 0xc420155368, 0x1, 0x1, 0x18, 0x0)
        /usr/local/go/src/flag/flag.go:814 +0x116
flag.(*FlagSet).parseOne(0xc420016ae0, 0x1, 0x0, 0x0)
        /usr/local/go/src/flag/flag.go:873 +0xad8
flag.(*FlagSet).Parse(0xc420016ae0, 0xc42000c180, 0x14, 0x14, 0x1578d78, 0x0)
        /usr/local/go/src/flag/flag.go:915 +0x60
github.com/vmware/vic/vendor/github.com/urfave/cli.Command.Run(0xd4144a, 0x6, 0x0, 0x0, 0x0, 0x0, 0x0, 0xd43045, 0xa, 0x0, ...)
        /opt/go/src/github.com/vmware/vic/vendor/github.com/urfave/cli/command.go:120 +0x69c
github.com/vmware/vic/vendor/github.com/urfave/cli.(*App).Run(0xc420194000, 0xc42000c160, 0x16, 0x16, 0x0, 0x0)
        /opt/go/src/github.com/vmware/vic/vendor/github.com/urfave/cli/app.go:236 +0x60c
main.main()
        /opt/go/src/github.com/vmware/vic/cmd/vic-machine/main.go:98 +0x1756
[vagrant@devbox:/opt/go/src/github.com/vmware/vic(master)]
```

to remove the script from the equation

```
[vagrant@devbox:/opt/go/src/github.com/vmware/vic(master)] /opt/go/src/github.com/vmware/vic/bin/vic-machine-linux create -t 10.118.69.193 -u root -p {pass} -n ZzZ -r /ha-datacenter/host/office2-sfo2-dhcp299.eng.vmware.com/Resources --image-datastore datastore1 --volume-store datastore1:default --container-network VM Network:external --timeout 10m --no-tls -f
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x5a3b08]

goroutine 1 [running]:
panic(0xb95100, 0xc4200120c0)
        /usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/vmware/vic/pkg/flags.(*optionalString).String(0xc4200280e0, 0xc22de0, 0xc4200280e0)
        /opt/go/src/github.com/vmware/vic/pkg/flags/optional_string.go:38 +0x8
flag.isZeroValue(0xc42001b340, 0xd40b9b, 0x5, 0xd53f73)
        /usr/local/go/src/flag/flag.go:393 +0x12d
flag.(*FlagSet).PrintDefaults.func1(0xc42001b340)
        /usr/local/go/src/flag/flag.go:467 +0x1d5
flag.(*FlagSet).VisitAll(0xc420016ae0, 0xc420157160)
        /usr/local/go/src/flag/flag.go:323 +0x67
flag.(*FlagSet).PrintDefaults(0xc420016ae0)
        /usr/local/go/src/flag/flag.go:476 +0x46
flag.defaultUsage(0xc420016ae0)
        /usr/local/go/src/flag/flag.go:509 +0x96
flag.(*FlagSet).usage(0xc420016ae0)
        /usr/local/go/src/flag/flag.go:825 +0x55
flag.(*FlagSet).failf(0xc420016ae0, 0xd6abd9, 0x22, 0xc420157368, 0x1, 0x1, 0x18, 0x0)
        /usr/local/go/src/flag/flag.go:814 +0x116
flag.(*FlagSet).parseOne(0xc420016ae0, 0x1, 0x0, 0x0)
        /usr/local/go/src/flag/flag.go:873 +0xad8
flag.(*FlagSet).Parse(0xc420016ae0, 0xc42000c1a0, 0x15, 0x16, 0x1578d78, 0x0)
        /usr/local/go/src/flag/flag.go:915 +0x60
github.com/vmware/vic/vendor/github.com/urfave/cli.Command.Run(0xd4144a, 0x6, 0x0, 0x0, 0x0, 0x0, 0x0, 0xd43045, 0xa, 0x0, ...)
        /opt/go/src/github.com/vmware/vic/vendor/github.com/urfave/cli/command.go:120 +0x69c
github.com/vmware/vic/vendor/github.com/urfave/cli.(*App).Run(0xc42000c300, 0xc42000c180, 0x17, 0x18, 0x0, 0x0)
        /opt/go/src/github.com/vmware/vic/vendor/github.com/urfave/cli/app.go:236 +0x60c
main.main()
        /opt/go/src/github.com/vmware/vic/cmd/vic-machine/main.go:98 +0x1756
[vagrant@devbox:/opt/go/src/github.com/vmware/vic(master)]
```

Yep, I just did

```
[vagrant@devbox:/opt/go/src/github.com/vmware/vic(1578)] bin/vic-machine-linux create -t vcva-clovervm.eng.vmware.com -u root -p 'PW' -n ZzZ -r /clovervm/host/bonneville/Resources  --image-datastore FC-shared01 --external-network DPortGroup  --bridge-network Bonneville --no-tls -f
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x5abc18]

goroutine 1 [running]:
panic(0xb9cf60, 0xc4200100d0)
        /usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/vmware/vic/pkg/flags.(*optionalString).String(0xc42002a0f0, 0xc2ac40, 0xc42002a0f0)
        /opt/go/src/github.com/vmware/vic/pkg/flags/optional_string.go:38 +0x8
flag.isZeroValue(0xc42001d2c0, 0xd48da8, 0x5, 0xd5c262)
        /usr/local/go/src/flag/flag.go:393 +0x12d
flag.(*FlagSet).PrintDefaults.func1(0xc42001d2c0)
        /usr/local/go/src/flag/flag.go:467 +0x1d5
flag.(*FlagSet).VisitAll(0xc420016b40, 0xc420154ee8)
        /usr/local/go/src/flag/flag.go:323 +0x67
flag.(*FlagSet).PrintDefaults(0xc420016b40)
        /usr/local/go/src/flag/flag.go:476 +0x46
flag.defaultUsage(0xc420016b40)
        /usr/local/go/src/flag/flag.go:509 +0x96
flag.(*FlagSet).usage(0xc420016b40)
        /usr/local/go/src/flag/flag.go:825 +0x55
flag.(*FlagSet).failf(0xc420016b40, 0xd72fcc, 0x22, 0xc4201550f0, 0x1, 0x1, 0x18, 0x0)
        /usr/local/go/src/flag/flag.go:814 +0x116
flag.(*FlagSet).parseOne(0xc420016b40, 0x1, 0x0, 0x0)
        /usr/local/go/src/flag/flag.go:873 +0xad8
flag.(*FlagSet).Parse(0xc420016b40, 0xc42000c160, 0x12, 0x12, 0x1583d38, 0x0)
        /usr/local/go/src/flag/flag.go:915 +0x60
github.com/vmware/vic/vendor/github.com/urfave/cli.Command.Run(0xd4965c, 0x6, 0x0, 0x0, 0x0, 0x0, 0x0, 0xd4b262, 0xa, 0x0, ...)
        /opt/go/src/github.com/vmware/vic/vendor/github.com/urfave/cli/command.go:120 +0x69c
github.com/vmware/vic/vendor/github.com/urfave/cli.(*App).Run(0xc420194000, 0xc42000c140, 0x14, 0x14, 0x0, 0x0)
        /opt/go/src/github.com/vmware/vic/vendor/github.com/urfave/cli/app.go:236 +0x60c
main.main()
        /opt/go/src/github.com/vmware/vic/cmd/vic-machine/main.go:106 +0x215e
[vagrant@devbox:/opt/go/src/github.com/vmware/vic(1578)] bin/vic-machine-linux create -t vcva-clovervm.eng.vmware.com -u root -p 'PW' -n ZzZ -r /clovervm/host/bonneville/Resources  --image-store FC-shared01 --external-network DPortGroup  --bridge-network Bonneville --no-tls -f
INFO[2016-08-17T00:09:09Z] ### Installing VCH ####
WARN[2016-08-17T00:09:09Z] Configuring without TLS - to enable drop --no-tls or use --key/--cert parameters
```

--- a/pkg/flags/optional_string.go
+++ b/pkg/flags/optional_string.go
@@ -35,7 +35,7 @@ func (b *optionalString) Get() interface{} {
 }
 
 func (b *optionalString) String() string {
-	if *b.val == nil {
+	if b.val == nil || *b.val == nil {
 		return "<nil>"
 	}
 	return **b.val

--- a/pkg/flags/shares_flag.go
+++ b/pkg/flags/shares_flag.go
@@ -48,7 +48,7 @@ func (s *ShareFlag) Set(val string) error {
 }
 
 func (s *ShareFlag) String() string {
-	if *s.shares == nil {
+	if s.shares == nil || *s.shares == nil {
 		return "<nil>"
 	}
 	switch (*s.shares).Level {

--- a/pkg/flags/url_flag.go
+++ b/pkg/flags/url_flag.go
@@ -46,7 +46,7 @@ func (f *URLFlag) Get() interface{} {
 }
 
 func (f *URLFlag) String() string {
-	if *f.u == nil {
+	if f.u == nil || *f.u == nil {
 		return "<nil>"
 	}
 	return (*f.u).String()

